kolibri/core/auth/serializers.py

```diff
@@ -63,6 +66,26 @@ def save(self, **kwargs):
 
     def validate(self, attrs):
         username = attrs.get("username")
+        username_validator = RegexValidator(
+            r'[\s`~!@#$%^&*()\-+={}\[\]\|\\\/:;"\'<>,\.\?]',
+            "Enter a valid username. This value can contain only letters, numbers, and underscores.",
+            code=error_constants.INVALID,
+            inverse_match=True,
+        )
+        try:
+            username_validator(username)
+        except DjangoValidationError as e:
+            raise serializers.ValidationError({"username": e.message})
+
+        try:
+            MaxLengthValidator(
+                30, "Required. 30 characters or fewer. Letters and digits only"
+            )(username)
+        except DjangoValidationError as e:
+            raise serializers.ValidationError(
+                {"username": e.message}, code=error_constants.MAX_LENGTH
+            )
+
```

Comment:
Oh, it looks like this method was never properly handling a PATCH where the username isn't specified (or it was, but kind of by accident?) - but this is why you have a failing test, because the `username` is returning `None` from the `.get` call, and so you are trying to run validation on `None`.

Reply:
yep, it was an easy fix though

Reply:
Oh, I'd think maybe we should skip username validation when it's not passed in, otherwise this will now throw errors if someone with an email username tries to edit their password?

Reply:
not sure, what  your proposal is the first thing I thought, but later in the validation function there's code that requires the username https://github.com/learningequality/kolibri/blob/release-v0.16.x/kolibri/core/auth/serializers.py#L79 
A simple if `username is not None` to validate username would be enough for this PR purposes but I'm not sure of the side effects it'll have to try something like 
`FacilityUser.objects.get(username__iexact=None, facility=facility)`
is there any chance to have a null username in some old db?

I'm happy to change it if you can confirm that's impossible.

Reply:
Yeah, I think just not running the username validation when we're not setting it is sufficient - and in that case we also don't need to do the DB request to check against other FacilityUsers, because the serializer doesn't allow us to change facilities, so we're only checking for a changed username clashing.

One last thought - in the frontend, should we prevent a username from being editable if it doesn't match our 30 character non-special char regex? If someone just tweaks it there, then it won't save.

Reply:
Applied a small change according to what we've commented here